Add support for authentication of polymorphic principals

evnm opened the pull request:
This change adds support for polymorphic `java.security.Principal` classes for use with Dropwizard's authentication mechanism. See #1318 for background.

My company has been running these custom `AuthValueFactoryProvider` and `AuthDynamicFeature` classes in our API server for months without issue. As stated in the ticket, we have a need to issue multiple classes of API keys to different API consumers and this change has suited the job well.

Note that the new `PolymorphicAuthDynamicFeature` only supports authentication, as opposed to authorization. [`AuthDynamicFeature`'s short-circuit on `RolesAllowed`/`DenyAll`/`PermitAll`](https://github.com/dropwizard/dropwizard/blob/2c3cca983f7b844a9f4bbaf93865120eae553e2a/dropwizard-auth/src/main/java/io/dropwizard/auth/AuthDynamicFeature.java#L39-L41) doesn't jive with per-Principal-subclass filtering, so I opted to limit the feature to authentication.

arteam commented on dropwizard-auth/src/main/java/io/dropwizard/auth/AuthValueFactoryProvider.java:
```diff
@@ -19,28 +16,31 @@
-     */
-    private final Class<T> principalClass;
+public class AuthValueFactoryProvider<T extends Principal>
+    extends PolymorphicAuthValueFactoryProvider
```
I would prefer `AuthValueFactoryProvider` not to be extended from `PolymorphicAuthValueFactoryProvider`.

This ties these classes rather highly without a need. I think we can find another ways to reuse the common code, if this is a concern.

arteam commented on dropwizard-auth/src/main/java/io/dropwizard/auth/PolymorphicAuthDynamicFeature.java:
```diff
@@ -0,0 +1,45 @@
+ * <p>Note that this feature only pertains to <i>authentication</i>.
+ * To enable authorization, use {@link AuthDynamicFeature}.</p>
+ */
+public class PolymorphicAuthDynamicFeature<T extends Principal> implements DynamicFeature {
```
Could you provide a minimal test case that uses this feature with several auth filters? I would prefer to see tests for every contributed functionality.

arteam commented on dropwizard-auth/src/main/java/io/dropwizard/auth/PolymorphicAuthValueFactoryProvider.java:
```diff
@@ -0,0 +1,133 @@
+@Singleton
+public class PolymorphicAuthValueFactoryProvider<T extends Principal>
+    extends AbstractValueFactoryProvider
+{
```
Nit: Please, remove a line break before the opening brace. In Dropwizard we follow [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style) on this matter.

arteam commented on dropwizard-auth/src/main/java/io/dropwizard/auth/PolymorphicAuthValueFactoryProvider.java:
```diff
@@ -0,0 +1,133 @@
+
+        @Override
+        protected void configure() {
+            bind(new PrincipalClassSetProvider<>(principalClassSet))
```
Nit: I think, it would be better to be consistent with `AuthValueFactoryProvider` and  have each bind statement (apart from the last) on one line.

arteam commented on dropwizard-auth/src/main/java/io/dropwizard/auth/PolymorphicAuthValueFactoryProvider.java:
```diff
@@ -0,0 +1,133 @@
+     *
+     * @param mpep                   multivalued parameter extractor provider
+     * @param injector               injector instance
+     * @param principalClassProvider provider of the principal class
```
The parameter name in the docs doesn't match the actual parameter name `principalClassSetProvider`

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/principal/NoAuthPrincipalEntityResource.java:
```diff
@@ -13,7 +13,9 @@
 import static org.assertj.core.api.Assertions.assertThat;
 
 /**
- * Contains resource methods which don't authenticate but use principal instance injection and thus might be affected by authentication logic.
```
Please, don't reformat source code without a need. It makes changes cleaner and focused.

arteam commented:
@evnm, thanks for your contribution! I reviewed it and this looks good in general. I've added some comments to the patch, see them inline.

evnm commented:
Thanks for the thorough review, @arteam. I've heeded your feedback and updated the diff.

In digging further into the dropwizard-auth tests, I decided to collapse `AuthBaseResourceConfig` into `AbstractAuthResourceConfig`. I don't think the usefulness of the two separate classes warrants the complexity imposed by a class hierarchy of `DropwizardResourceConfig` within a test package. So, this change now comprises two commits, the first of which does a bit of cleanup within the auth tests.

The meat of the change which adds support for polymorphic `Principal`s now reflects your comments. `PolymorphicAuthValueFactoryProvider` and `AuthValueFactoryProvider` each extend `AbstractValueFactoryProvider` directly. The meat of their respective `createValueFactory` methods is offloaded to a package-private `PrincipalContainerRequestValueFactory` class. The rest of the two provider classes is basically boilerplate.

And I added a `PolymorphicPrincipalEntityTest` which wires the new pieces up and ensures that authentication by multiple Principal subclasses in a resource actually works.

I'm not sure how (or if it's even worth it) to test `PrincipalContainerRequestValueFactory`. Taking a look at Jersey's source, `AbstractContainerRequestValueFactory` itself isn't tested, so I couldn't find a precedent for mocking `ContainerRequest` injection functionality.

Still not a fan of the "PolymorphicFoo" naming convention, but I still can't come up with anything better.

arteam commented on dropwizard-auth/src/main/java/io/dropwizard/auth/PrincipalContainerRequestValueFactory.java:
```diff
@@ -0,0 +1,23 @@
+
+/**
+ * A value factory which extracts the {@link Principal} from the
+ * current {@link ContainerRequest} instance.
```
Need to provide `import org.glassfish.jersey.server.ContainerRequest;` to use the short name.

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/basic/BasicAuthProviderTest.java:
```diff
@@ -8,8 +8,13 @@
-        protected ContainerRequestFilter getAuthFilter() {
+    public static class BasicAuthTestResourceConfig extends AbstractAuthResourceConfig {
+        public BasicAuthTestResourceConfig() {
+            super();
```
I believe, `super()` is redundant, because the Java compiler automatically inserts a call to the default constructor of a superclass.

evnm replied:
You're right. I got a bit overzealous with these `super()`s while trying to get resource injection to compile with the new class hierarchy. Will remove these.

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/basic/BasicCustomAuthProviderTest.java:
```diff
@@ -9,8 +9,13 @@
-        protected ContainerRequestFilter getAuthFilter() {
+    public static class BasicAuthTestResourceConfig extends AbstractAuthResourceConfig {
+        public BasicAuthTestResourceConfig() {
+            super();
```
Ditto for `super()`

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/oauth/OAuthCustomProviderTest.java:
```diff
@@ -5,25 +5,30 @@
+public class OAuthCustomProviderTest extends AuthBaseTest<OAuthCustomProviderTest.OAuthTestResourceConfig> {
+    public static class OAuthTestResourceConfig extends AbstractAuthResourceConfig {
+        public OAuthTestResourceConfig() {
+            super();
```
Ditto for `super()`

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/oauth/OAuthProviderTest.java:
```diff
@@ -5,9 +5,14 @@
+public class OAuthProviderTest extends AuthBaseTest<OAuthProviderTest.OAuthTestResourceConfig>{
+    public static class OAuthTestResourceConfig extends AbstractAuthResourceConfig {
+        public OAuthTestResourceConfig() {
+            super();
```
Ditto for `super()`

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/principal/NoAuthPrincipalEntityTest.java:
```diff
@@ -41,6 +45,7 @@ protected DeploymentContext configureDeployment() {
     public static class NoAuthPrincipalInjectedResourceConfig extends AbstractAuthResourceConfig {
 
         public NoAuthPrincipalInjectedResourceConfig() {
+            super();
```
Ditto for `super()`

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/principal/PolymorphicPrincipalEntityResource.java:
```diff
@@ -0,0 +1,39 @@
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import java.lang.annotation.ElementType;
```
Redundant imports of the `java.lang.annotation` package

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/principal/PolymorphicPrincipalEntityTest.java:
```diff
@@ -0,0 +1,150 @@
+        }
+
+        @Override protected AbstractBinder getAuthBinder() {
+            final Set<Class<? extends Principal>> principalClassSet = new HashSet<>();
```
I think using `ImmutableSet.of(JsonPrincipal.class, NullPrincipal.class)` here would be more terse.

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/principal/PolymorphicPrincipalEntityTest.java:
```diff
@@ -0,0 +1,150 @@
+            final Set<Class<? extends Principal>> principalClassSet = new HashSet<>();
+            principalClassSet.add(JsonPrincipal.class);
+            principalClassSet.add(NullPrincipal.class);
+            return new PolymorphicAuthValueFactoryProvider.Binder<Principal>(principalClassSet);
```
We could just use the diamond type here instead explicit type argument.

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/principal/PolymorphicPrincipalEntityTest.java:
```diff
@@ -0,0 +1,150 @@
+        @Override protected DynamicFeature getAuthDynamicFeature(ContainerRequestFilter authFilter) {
+            final Authenticator<BasicCredentials, JsonPrincipal> jsonAuthenticator = credentials -> {
+                if (credentials.getUsername().equals(JSON_USERNAME)) {
+                    return Optional.<JsonPrincipal>of(new JsonPrincipal(credentials.getUsername()));
```
Since Java 8 the compiler can infer such explicit type arguments, so `<JsonPrincipal>` is redundant, I beleive.

evnm replied:
You're right. Another vestige of a battle with the compiler. Will remove.

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/principal/PolymorphicPrincipalEntityTest.java:
```diff
@@ -0,0 +1,150 @@
+
+            final Authenticator<BasicCredentials, NullPrincipal> nullAuthenticator = credentials -> {
+                if (credentials.getUsername().equals(NULL_USERNAME)) {
+                    return Optional.<NullPrincipal>of(new NullPrincipal());
```
Ditto for `<NullPrincipal>`

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/principal/PolymorphicPrincipalEntityTest.java:
```diff
@@ -0,0 +1,150 @@
+                .setAuthenticator(nullAuthenticator)
+                .buildAuthFilter();
+
+            return new PolymorphicAuthDynamicFeature(ImmutableMap.of(
```
To avoid a compiler warning about an unchecked call, we should provide type argument `<Principal>` for `PolymorphicAuthDynamicFeature`.

arteam commented on dropwizard-auth/src/test/java/io/dropwizard/auth/principal/NoAuthPolymorphicPrincipalEntityTest.java:
```diff
@@ -0,0 +1,128 @@
+        }
+
+        @Override protected AbstractBinder getAuthBinder() {
+            final Set<Class<? extends PrincipalImpl>> principalClassSet = new HashSet<>();
```
Ditto. I think using `ImmutableSet.of(...)` is a more terse and clearer.

arteam commented:
Thanks, Evan. Your reasons about collapsing `AuthBaseResourceConfig` and introdcing `PrincipalContainerRequestValueFactory` seem sensible for me. I've added comments about minor details of the implementation, but overall this is ready to merge.

Also, could you be so kind to update the [release notes](https://github.com/dropwizard/dropwizard/blob/master/docs/source/about/release-notes.rst) for 1.0 version? This would make easier for users
to track this improvement.

evnm commented:
Change should now be good to go. I took the liberty of adding myself to the contributors page, too. Hopefully not too gauche.

Thanks again for the thoughtful review, @arteam!

arteam commented:
Thanks for contributing to Dropwizard!

> Hopefully not too gauche.

Not at all. But we usually do this automatically with a [Ruby script|https://github.com/dropwizard/dropwizard/blob/master/docs/list_contributors.rb]. We run it time to time to actualize the list.

BenRomberg commented:
Don't know if this is the right place to say "thank you", but this contribution just saved me a lot of work! Thanks a lot @evnm @arteam =)
